### PR TITLE
Feature bash completion

### PR DESCRIPTION
**Pull request by userdocs**
## Description

Provides bash completion for `box` commands.

Some are statically defined like `panel fix`

`install` `remove` and `upgrade` are dynamic and will return results based on the files that exist relative to those commands.

It is pretty easy to expand or adapt if needed.

## Fixes issues: 
- na

## Proposed Changes:
- tabbed `box` completion

## Change Categories
- New feature 

## Checklist

- [ ] Docs have been made OR are not necessary
    - PR link: 
- [ ] Changes to panel have been made OR are not necessary
    - PR link: 
- [ ] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [ ] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [ ] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [ ] I have commented my code, particularly in hard-to-understand areas

### Architectures

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|			|			|			|				|
| Bionic		|			|			|			|				|
| Buster		|			|			|			|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### CODE REVIEW

File: scripts/update/completion.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+#
+# Set the correct permissions for the completion file and then create a symlink to the bash completion file.
+chmod 644 '/etc/swizzin/sources/bash_completion.d/swizzin'
+ln -fs '/etc/swizzin/sources/bash_completion.d/swizzin' '/etc/bash_completion.d/swizzin'
```
Comment (flying-sausages) [inline scripts/update/completion.sh:5]:
```suggestion
if [ ! -e /etc/bash_completion.d/swizzin ]; then
    chmod 644 '/etc/swizzin/sources/bash_completion.d/swizzin'
    ln -s '/etc/swizzin/sources/bash_completion.d/swizzin' '/etc/bash_completion.d/swizzin'
    echo_info "Shell completions for the box command have been installed, try pressing <tab> the next time you type 'box'" 
fi
```

File: setup.sh [hunk 2]
```diff
@@ -375,6 +375,9 @@ function _post() {
         echo_info "You can now use the box command to manage swizzin features, e.g. \`box install nginx panel\`"
     fi
     echo_docs getting-started/box-basics
+    # Set the correct permissions for the completion file and then create a symlink to the bash completion file.
+    chmod 644 '/etc/swizzin/sources/bash_completion.d/swizzin'
+    ln -fs '/etc/swizzin/sources/bash_completion.d/swizzin' '/etc/bash_completion.d/swizzin'
```
Comment (flying-sausages) [inline setup.sh:380]: I don't think we need to abuse the `-f` flag here
```suggestion
    ln -s '/etc/swizzin/sources/bash_completion.d/swizzin' '/etc/bash_completion.d/swizzin'
```

Comment (userdocs) [reply setup.sh:380]: i don't think abuse is correct. no different from `cp` or `mv` or `rm. just a forces the outcome. no need to check it it exists. just create it.

File: sources/bash_completion.d/swizzin [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+_command_complete() {
+    local cur prev
+    cur=${COMP_WORDS[COMP_CWORD]}
+    prev=${COMP_WORDS[COMP_CWORD - 1]}
+    readarray -d ' ' apps_array < <(find "/etc/swizzin/scripts/install/" -maxdepth 1 -mindepth 1 -type f -printf "%f ")
+    readarray -d ' ' upgrade_array < <(find "/etc/swizzin/scripts/upgrade/" -maxdepth 1 -mindepth 1 -type f -printf "%f ")
+
+    case ${COMP_CWORD} in
+        1)
+            mapfile -t COMPREPLY < <(compgen -W "install remove adduser deluser chpasswd update upgrade panel list" -- "${cur}")
```
Comment (flying-sausages) [inline sources/bash_completion.d/swizzin:14]: Might be nice to organise this by "topic" a little, and added some others left.
```suggestion
            mapfile -t COMPREPLY < <(compgen -W "install remove upgrade update adduser deluser chpasswd  panel rtx rmgrsec list help" -- "${cur}")
```

Comment (userdocs) [reply sources/bash_completion.d/swizzin:14]: compgen will sort it automatically.

File: scripts/update/completion.sh (hunk 1 above)
Comment (userdocs) [reply scripts/update/completion.sh:5]: I don't think there is a need to check it exists, it's a symlink. just recreate it and we are done

File: sources/bash_completion.d/swizzin [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+#
+# ln -fs /etc/swizzin/sources/bash_completion.d/swizzin /etc/bash_completion.d/swizzin
+#
+_command_complete() {
+    local cur prev
+    cur=${COMP_WORDS[COMP_CWORD]}
+    prev=${COMP_WORDS[COMP_CWORD - 1]}
+    readarray -d ' ' apps_array < <(find "/etc/swizzin/scripts/install/" -maxdepth 1 -mindepth 1 -type f -printf "%f ")
```
Comment (flying-sausages) [inline sources/bash_completion.d/swizzin:9]: What if we remove things that are already installed from this, i.e. remove from this array based on what's in in `/install`? Is it worth the trouble?

File: sources/bash_completion.d/swizzin [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        1)
+            mapfile -t COMPREPLY < <(compgen -W "install remove adduser deluser chpasswd update upgrade panel list" -- "${cur}")
+            ;;
+        2)
+            case ${prev} in
+                install)
+                    mapfile -t COMPREPLY < <(compgen -W "$(printf "%s " "${apps_array[@]%.*}")" -- "${cur}")
+                    ;;
+                remove)
+                    mapfile -t COMPREPLY < <(compgen -W "$(printf "%s " "${apps_array[@]%.*}")" -- "${cur}")
```
Comment (flying-sausages) [inline sources/bash_completion.d/swizzin:22]: i feel like this should be based off what's in `/install`

File: sources/bash_completion.d/swizzin (hunk 3 above)
Comment (flying-sausages) [reply sources/bash_completion.d/swizzin:14]: Yeah, the order is lost, shoud still add the rest of the missing commands. I changed the order back to just show the additions

**Review by flying-sausages — APPROVED** [review]
Merge when ready

**Pull request merged**